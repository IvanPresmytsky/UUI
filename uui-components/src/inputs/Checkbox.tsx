import * as React from 'react';
import { cx, IHasTabIndex, uuiMarkers } from '@epam/uui-core';
import css from './Checkbox.module.scss';
import {
    Icon, uuiMod, uuiElement, isEventTargetInsideClickable, CheckboxCoreProps, UuiContexts, UuiContext,
} from '@epam/uui-core';
import { IconContainer } from '../layout/IconContainer';

export interface CheckboxProps extends CheckboxCoreProps, IHasTabIndex {
    /** Render callback for checkbox label.
     * If omitted, 'label' prop value will be rendered.
     */
    renderLabel?(): React.ReactNode;

    /** ID provided to the 'input' node */
    id?: string;

    /** Check icon.
     * Usually it has a default implementation in skins, so providing this is only necessary if you want to replace the default icon
     */

    icon?: Icon;
    /** Indeterminate state icon.
     * Usually it has a default implementation in skins, so providing this is only necessary if you want to replace the default icon
     *  */
    indeterminateIcon?: Icon;
}

export class Checkbox extends React.Component<CheckboxProps> {
    static contextType = UuiContext;
    context: UuiContexts;

    handleChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
        !isEventTargetInsideClickable(e) && this.props.onValueChange(!this.props.value);

        if (this.props.getValueChangeAnalyticsEvent) {
            const event = this.props.getValueChangeAnalyticsEvent(!this.props.value, this.props.value);
            this.context.uuiAnalytics.sendEvent(event);
        }
    };

    handleAriaCheckedValue = (indeterminate: boolean, value: boolean): boolean | 'mixed' => {
        if (indeterminate) {
            return 'mixed';
        }

        return value == null ? false : value;
    };

    render() {
        let label = this.props.label;
        if (this.props.renderLabel) {
            label = this.props.renderLabel();
        }
        const ariaCheckedValue = this.handleAriaCheckedValue(this.props.indeterminate, this.props.value);

        return (
            <label
                className={ cx(
                    css.container,
                    uuiElement.checkboxContainer,
                    this.props.cx,
                    this.props.isDisabled && uuiMod.disabled,
                    this.props.isReadonly && uuiMod.readonly,
                    this.props.isInvalid && uuiMod.invalid,
                    !this.props.isReadonly && !this.props.isDisabled && uuiMarkers.clickable,
                ) }
                ref={ this.props.forwardedRef }
                { ...this.props.rawProps }
            >
                <div
                    className={ cx(uuiElement.checkbox, (this.props.value || this.props.indeterminate) && uuiMod.checked) }
                    onFocus={ this.props.onFocus }
                    onBlur={ this.props.onBlur }
                >
                    <input
                        type="checkbox"
                        onChange={ !this.props.isReadonly ? this.handleChange : undefined }
                        disabled={ this.props.isDisabled }
                        aria-disabled={ this.props.isDisabled || undefined }
                        readOnly={ this.props.isReadonly }
                        aria-readonly={ this.props.isReadonly || undefined }
                        checked={ this.props.value || false }
                        aria-checked={ ariaCheckedValue }
                        required={ this.props.isRequired }
                        aria-required={ this.props.isRequired || undefined }
                        tabIndex={ this.props.tabIndex || this.props.isReadonly || this.props.isDisabled ? -1 : 0 }
                        id={ this.props.id }
                    />
                    { this.props.value && !this.props.indeterminate && <IconContainer icon={ this.props.icon } /> }
                    { this.props.indeterminate && <IconContainer icon={ this.props.indeterminateIcon } /> }
                </div>
                { label && <div className={ uuiElement.inputLabel }>{ label }</div> }
            </label>
        );
    }
}
