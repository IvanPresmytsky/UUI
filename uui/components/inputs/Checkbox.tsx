import { withMods } from '@epam/uui-core';
import * as uuiComponents from '@epam/uui-components';
import { ReactComponent as Check } from '../../icons/check-18.svg';
import { ReactComponent as PartlySelect } from '../../icons/partly-select-18.svg';
import css from './Checkbox.module.scss';

interface CheckboxMods {
    /**
     * Defines component size.
     * @default '18'
     */
    size?: '12' | '18';
    /**
     * Defines the different edit modes.
     * @default 'form'
     */
    mode?: 'form' | 'cell';
}

/** Represents the properties of the Checkbox component. */
export type CheckboxProps = CheckboxMods & uuiComponents.CheckboxProps;

function applyCheckboxMods(mods: CheckboxMods) {
    return [
        css.root,
        css['size-' + (mods.size || '18')],
        css['mode-' + (mods.mode || 'form')],
        'uui-color-primary',
    ];
}

const applyUUICheckboxProps = (props: CheckboxProps) => {
    const defaultIcon = Check;
    const defaultIndeterminateIcon = PartlySelect;
    return {
        icon: props.icon ? props.icon : defaultIcon,
        indeterminateIcon: props.indeterminateIcon ? props.indeterminateIcon : defaultIndeterminateIcon,
    };
};

export const Checkbox = withMods<uuiComponents.CheckboxProps, CheckboxMods>(uuiComponents.Checkbox, applyCheckboxMods, applyUUICheckboxProps);
