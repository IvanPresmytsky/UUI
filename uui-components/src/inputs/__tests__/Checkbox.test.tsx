import React from 'react';
import { Checkbox, CheckboxProps } from '../Checkbox';
import { render, screen, fireEvent, setupComponentForTest, userEvent } from '@epam/uui-test-utils';

async function setupCheckbox(params: Partial<CheckboxProps>) {
    const { mocks, setProps } = await setupComponentForTest<CheckboxProps>(
        (context) => ({
            value: params.value,
            onValueChange: params.onValueChange || jest.fn().mockImplementation((newValue) => {
                context.current.setProperty('value', newValue);
            }),
            ...params,
        }),
        (props) => {
            return <Checkbox { ...props } />;
        },
    );
    return {
        setProps,
        mocks,
    };
}

describe('Checkbox', () => {
    it('should handle change event', async () => {
        const onValueChange = jest.fn();
        const getValueChangeAnalyticsEvent = jest.fn();
        await setupCheckbox({
            value: false,
            onValueChange,
            getValueChangeAnalyticsEvent,
        });
        const input = screen.getByRole('checkbox');

        fireEvent.click(input);

        expect(onValueChange).toHaveBeenCalledWith(true);
        expect(getValueChangeAnalyticsEvent).toHaveBeenCalled();
    });

    it('should handle keyboard click', async () => {
        const onValueChange = jest.fn();
        const getValueChangeAnalyticsEvent = jest.fn();
        await setupCheckbox({
            value: false,
            onValueChange,
            getValueChangeAnalyticsEvent,
        });
        const input = screen.getByRole('checkbox');

        await userEvent.type(input, '{space}');

        expect(onValueChange).toHaveBeenCalledWith(true);
        expect(getValueChangeAnalyticsEvent).toHaveBeenCalled();
    });

    it('should handle label click', async () => {
        const onValueChange = jest.fn();
        const getValueChangeAnalyticsEvent = jest.fn();
        await setupCheckbox({
            value: false,
            onValueChange,
            getValueChangeAnalyticsEvent,
            label: 'Label',
        });
        const label = screen.getByText('Label');

        label.click();

        expect(onValueChange).toHaveBeenCalledWith(true);
        expect(getValueChangeAnalyticsEvent).toHaveBeenCalled();
    });

    it('should not handle change event when readonly', () => {
        const onValueChange = jest.fn();
        render(<Checkbox value={ false } onValueChange={ onValueChange } isReadonly />);
        const input = screen.getByRole('checkbox');

        fireEvent.click(input);
        expect(onValueChange).not.toHaveBeenCalled();
    });

    it('should handle focus event', () => {
        const onFocus = jest.fn();
        render(<Checkbox value={ false } onValueChange={ jest.fn } onFocus={ onFocus } />);
        const input = screen.getByRole('checkbox');

        input.focus();

        expect(onFocus).toHaveBeenCalled();
        expect(input).toHaveFocus();
    });

    it('should handle blur event', () => {
        const onBlur = jest.fn();
        render(<Checkbox value={ false } onValueChange={ jest.fn } onBlur={ onBlur } />);
        const input = screen.getByRole('checkbox');

        input.focus();
        expect(input).toHaveFocus();
        input.blur();

        expect(onBlur).toHaveBeenCalled();
        expect(input).not.toHaveFocus();
    });
});
