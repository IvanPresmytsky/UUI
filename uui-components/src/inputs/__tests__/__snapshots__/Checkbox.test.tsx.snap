// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should be rendered correctly with disabled, invalid, readonly and required state 1`] = `
<label
  className="container uui-checkbox-container uui-disabled uui-readonly uui-invalid"
>
  <div
    className="uui-checkbox uui-checked"
  >
    <input
      aria-checked="mixed"
      aria-disabled={true}
      aria-readonly={true}
      aria-required={true}
      checked={true}
      disabled={true}
      readOnly={true}
      required={true}
      tabIndex={-1}
      type="checkbox"
    />
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    />
  </div>
  <div
    className="uui-input-label"
  >
    Test label
  </div>
</label>
`;

exports[`Checkbox when state equals isInvalid: true, Checkbox_container must have a 'uui-invalid' class 1`] = `
<label
  className="container uui-checkbox-container uui-disabled uui-readonly uui-invalid"
>
  <div
    className="uui-checkbox"
  >
    <input
      aria-checked={false}
      aria-disabled={true}
      aria-readonly={true}
      aria-required={true}
      checked={false}
      disabled={true}
      readOnly={true}
      required={true}
      tabIndex={-1}
      type="checkbox"
    />
  </div>
</label>
`;
